subsys/bluetooth/mesh/gatt_cli.c

```diff
@@ -0,0 +1,370 @@
+{
+	for (int i = 0; i < ARRAY_SIZE(servers); i++) {
+		if (servers[i].conn == conn) {
+			return &servers[i];
+		}
+	}
+
+	return NULL;
+}
+
+static uint8_t notify_func(struct bt_conn *conn,
+			   struct bt_gatt_subscribe_params *params,
+			   const void *data, uint16_t length)
+{
+	const uint8_t *val = data;
+	struct bt_mesh_gatt_server *server;
+
+	if (!data) {
+		BT_WARN("[UNSUBSCRIBED]");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+
+	server = find_server(conn);
+	if (!server) {
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (length < 1) {
+		BT_WARN("Too small Proxy PDU");
+		return BT_GATT_ITER_STOP;
+	}
+
+	(void)bt_mesh_proxy_msg_recv(conn, val, length);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t discover_func(struct bt_conn *conn,
+			     const struct bt_gatt_attr *attr,
+			     struct bt_gatt_discover_params *params)
+{
+	int err;
+	struct bt_mesh_gatt_server *server;
+
+	if (!attr) {
+		BT_DBG("GATT Services Discover complete");
+		(void)memset(params, 0, sizeof(*params));
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("[ATTRIBUTE] handle %u", attr->handle);
+
+	server = find_server(conn);
+	if (!server) {
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (!bt_uuid_cmp(server->discover.uuid, &server->gatt->srv_uuid.uuid)) {
+		server->svc_start_handle = attr->handle;
+
+		server->discover.uuid = &server->gatt->data_in_uuid.uuid;
+		server->discover.start_handle = attr->handle + 1;
+		server->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+
+		err = bt_gatt_discover(conn, &server->discover);
+		if (err) {
+			BT_DBG("Discover GATT data in char failed (err %d)", err);
+		}
+	} else if (!bt_uuid_cmp(server->discover.uuid,
+				&server->gatt->data_in_uuid.uuid)) {
+		server->data_in_handle = attr->handle + 1;
+
+		server->discover.uuid = &server->gatt->data_out_uuid.uuid;
+		server->discover.start_handle = server->svc_start_handle + 1;
+		server->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+
+		err = bt_gatt_discover(conn, &server->discover);
+		if (err) {
+			BT_DBG("Discover GATT data in char failed (err %d)", err);
+		}
+	} else if (!bt_uuid_cmp(server->discover.uuid,
+				&server->gatt->data_out_uuid.uuid)) {
+		server->discover.uuid = &server->gatt->data_out_cccd_uuid.uuid;
+		server->discover.start_handle = attr->handle + 2;
+		server->discover.type = BT_GATT_DISCOVER_DESCRIPTOR;
+		server->subscribe.value_handle = bt_gatt_attr_value_handle(attr);
+
+		err = bt_gatt_discover(conn, &server->discover);
+		if (err) {
+			BT_DBG("Discover GATT CCCD failed (err %d)", err);
+		}
+	} else {
+		server->subscribe.notify = notify_func;
+		server->subscribe.value = BT_GATT_CCC_NOTIFY;
+		server->subscribe.ccc_handle = attr->handle;
+
+		err = bt_gatt_subscribe(conn, &server->subscribe);
+		if (err && err != -EALREADY) {
+			BT_DBG("Subscribe failed (err %d)", err);
+		} else {
+			BT_DBG("[SUBSCRIBED]");
+
+			server->gatt->link_open(conn);
+		}
+
+		return BT_GATT_ITER_STOP;
```

Comment:
Is the order of the characteristics guaranteed by spec?

Reply:
It does not rely on the order of characteristics, I look for it from the service.